dicons.c

```diff
@@ -42,16 +43,32 @@ void append_row_from_file(GtkListStore *store, GFile *file)
 {
     GtkTreeIter iter;
     GFileInfo *file_info;
-    const gchar *display_name;
-    GdkPixbuf *pixbuf;
+    const gchar *display_name = NULL;
+    GdkPixbuf *pixbuf = NULL;
 
+    GKeyFile *keyfile = g_key_file_new ();
     file_info = g_file_query_info(file, "standard::*,ownser::user", 0, 0, 0);
-    display_name = g_file_info_get_display_name(file_info);
-    pixbuf = gtk_icon_info_load_icon(
-            gtk_icon_theme_lookup_by_gicon(
-                theme,
-                g_file_info_get_icon(file_info),
-                48, 0), 0);
+    if (g_key_file_load_from_file (keyfile, g_file_get_path(file), G_KEY_FILE_NONE, NULL))
+    {
+        GAppInfo* app = (GAppInfo*)g_desktop_app_info_new_from_keyfile (keyfile);
+        if(app) {
+            display_name = g_app_info_get_display_name(app);
+            pixbuf = gtk_icon_info_load_icon(
+                    gtk_icon_theme_lookup_by_gicon(
+                        theme,
+                        g_app_info_get_icon(app),
+                        48, 0), 0);
+        }
+    }
+
+    if(!display_name)
+        display_name = g_file_info_get_display_name(file_info);
+    if(!pixbuf)
+        pixbuf = gtk_icon_info_load_icon(
+                gtk_icon_theme_lookup_by_gicon(
+                    theme,
+                    g_file_info_get_icon(file_info),
+                    48, 0), 0);
```

Comment:
reduce code-duplication of `gtk_icon_info_load_icon`. Create a local variable `icon` of type `GIcon`. Set it to the app-icon or file-icon then you only have to write  `gtk_icon_info_load_icon([...], icon)` once